cumulusci/robotframework/Salesforce.py

```diff
@@ -499,6 +525,115 @@ def salesforce_insert(self, obj_name, **kwargs):
+                | ...  Age={{number}}
+
+            Which would generate:
+
+                | [{'FirstName': 'User 0', 'LastName': 'u4PNlGUD', 'type': 'Contact', 'Age': '0'},
+                |  {'FirstName': 'User 1', 'LastName': 'Kyd4PC5x', 'type': 'Contact', 'Age': '1'},
+                |  {'FirstName': 'User 2', 'LastName': 'n4EGs4Vp', 'type': 'Contact', 'Age': '2'}]
+
+            Jinja2 Expression Syntax is allowed so computed templates like this are also allowed: {{1000 + number}}
+
+            Templates can also be based on faker patterns like those described here:
+
+            https://faker.readthedocs.io/en/master/providers.html
+
+            Most examples can be pasted into templates verbatim:
+
+        @{objects}=  Salesforce Collection Generate  Contact  200
+        ...  FirstName={{fake.first_name}}
+        ...  LastName={{fake.last_name}}
+        ...  MailingStreet={{fake.street_address}}
+        ...  MailingCity=New York
+        ...  MailingState=NY
+        ...  MailingPostalCode=12345
+        ...  Email={{fake.email(domain="salesforce.com")}}
+
+           """
+        objs = []
+
+        def format_str(value, i):
+            if isinstance(value, str):
+                value = Template(value).render(
+                    number=i,
+                    random_str=self.random_string_generator,
+                    fake=self.faker_template_library,
+                )
+
+            return value
+
+        for i in range(int(number_to_create)):
+            formatted_fields = {
+                name: format_str(value, i) for name, value in fields.items()
+            }
+            newobj = self._salesforce_generate_object(obj_name, **formatted_fields)
+            objs.append(newobj)
+
+        return objs
+
+    def salesforce_collection_insert(self, objects):
+        """Inserts up to 200 records that were created with Salesforce Collection Generate.
+           The 200 record limit is enforced by the Salesforce APIs"""
+        assert (
+            not obj.get("id", None) for obj in objects
+        ), "Insertable objects should not have IDs"
+        assert (
+            len(objects) <= 200
+        ), "Cannot insert more than 200 objects with this keyword"
+
+        records = self.cumulusci.sf.restful(
+            "composite/sobjects",
+            method="POST",
+            json={"allOrNone": True, "records": objects},
+        )
+
+        for record, obj in zip(records, objects):
+            if record["errors"]:
+                raise AssertionError(*record["errors"])
+            self.store_session_record(obj["attributes"]["type"], record["id"])
+            obj["id"] = record["id"]
+            obj[STATUS_KEY] = record
+
+        return objects
+
+    def salesforce_collection_update(self, objects):
+        """Updates up to 200 records described as Robot/Python dictionaries"""
+        for obj in objects:
+            assert obj["id"], "Should be a list of objects with Ids"
```

Comment:
If you had 199 objects with ids and one without, this error message would be a bit misleading. You might want to make this message more specific, such as `Object number <n> does not have an 'id' property` or something like that.

Reply:
My thinking on this one is that if you have even a single object without an "id" then it almost certainly means you generated the objects in some non-standard way. You're not supposed to be generating these lists "by hand". These are supposed to be the return value from salesforce_collection_insert.

I will clarify the error message but more in the direction I'm describing.